pkg/workspace/apply.go

```diff
@@ -0,0 +1,98 @@
+package workspace
+
+import (
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/names"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	volumeNameBase = "ws"
+)
+
+// GetVolumes will return a dictionary where the keys are the names fo the workspaces bound in
+// wb and the value is the Volume to use. If the same Volume is bound twice, the resulting volumes
+// will both have the same name to prevent the same Volume from being attached to pod twice.
+func GetVolumes(wb []v1alpha1.WorkspaceBinding) map[string]corev1.Volume {
+	pvcs := map[string]corev1.Volume{}
+	v := map[string]corev1.Volume{}
+	for _, w := range wb {
+		name := names.SimpleNameGenerator.RestrictLengthWithRandomSuffix(volumeNameBase)
+		if w.PersistentVolumeClaim != nil {
+			// If it's a PVC, we need to check if we've encountered it before so we avoid mounting it twice
+			if vv, ok := pvcs[w.PersistentVolumeClaim.ClaimName]; ok {
+				v[w.Name] = vv
+			} else {
+				pvc := *w.PersistentVolumeClaim
+				v[w.Name] = corev1.Volume{
+					Name: name,
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &pvc,
+					},
+				}
+				pvcs[pvc.ClaimName] = v[w.Name]
+			}
+		} else if w.EmptyDir != nil {
+			ed := *w.EmptyDir
+			v[w.Name] = corev1.Volume{
+				Name: name,
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &ed,
+				},
+			}
+		}
+	}
+	return v
+}
+
+func getDeclaredWorkspace(name string, w []v1alpha1.WorkspaceDeclaration) (*v1alpha1.WorkspaceDeclaration, error) {
+	for _, workspace := range w {
+		if workspace.Name == name {
+			return &workspace, nil
+		}
+	}
+	// Trusting validation to ensure
+	return nil, fmt.Errorf("even though validation should have caught it, bound workspace %s did not exist in declared workspaces", name)
```

Comment:
nit: the bit about validation catching it might be superfluous info for the non-dev user?  or maybe not?  not sure, but also it's only a nit.  feel free to ignore.

Reply:
i agree but also it _shouldnt_ happen so i feel like this is extra information to indicate that something has gone horribly wrong? i dunno, i wish it didnt have to be there at all :S